docs/src/apalache/fold.md

```diff
@@ -0,0 +1,213 @@
+Folding in a nutshell
+==========================
+
+Apalache natively implements two operators users might be familiar with from the community modules or functional programming. Those operators are `FoldSet` and `FoldSeq`. This brief introduction to fold operators highlights the following:
+
+  1. What are the semantics of fold operators?
+  1. How do I use these operators in Apalache?
+  1. Should I use folding or recursion?
+  1. Examples of common operators defined with folds
+
+### Semantics of fold operators
+
+Folding refers to iterative application of a binary operator over a collection. Given an operator `Op`, a base value `b` and a collection of values `C`, folding `Op` over `C` starting with `b` is defined as follows:
+
+  1. If `C` is empty, the resulting value is `b`
+  1. If `C` is nonempty, we select an element `x` from `C`. If `C` is a sequence, `x` is the head value, and if `C` is a set, `x` is an arbitrary member. Then, we compute `b2 = Op(b, x)` and take `C2 = C without x` (`Tail(C)` or `C \ {x}` respectively). We continue the process, by computing the fold of `Op` over `C2` starting with `b2`.
+
+Note that this is the definition of a _left fold_ in the literature. Apalache does not implement a right fold.
```

Comment:
Not really, `FoldRight(Op, b, C) = FoldLeft(  Op*, b*, C reversed)` where `Op(a,b) = Op*(b,a)` and `b*` depends on the details (but likely `b`)

Reply:
Sure, you don't need both for expressivity. But I think that reversing a sequence would produce `O(n^2)` constraints in our encoding. Btw, the community modules have both left and right folds for sequences: https://github.com/tlaplus/CommunityModules/blob/master/modules/SequencesExt.tla

I would prefer having both left and right folds for sequences in `Apalache.tla`, even if the right fold is `foldl(rev(l))`. We could find a better encoding for it in the future.

Reply:
We could have both, in which case it would make more sense to have separate rewriting rules and operators at the IR level. However, this is definitely outside the scope of this tutorial (at least v1), since we don't have a right fold atm.